packages/apify/src/internals/charging.ts

```diff
@@ -0,0 +1,281 @@
+    private chargingState: Record<string, ChargingStateItem> | undefined = undefined;
+    private chargingLogDataset: Dataset | undefined;
+
+    private apifyClient: ApifyClient;
+
+    constructor(configuration: Configuration, apifyClient: ApifyClient) {
+        this.maxTotalChargeUsd = configuration.get('maxTotalChargeUsd') ?? Infinity;
+        this.isAtHome = configuration.get('isAtHome');
+        this.actorRunId = configuration.get('actorRunId');
+        this.purgeChargingLogDataset = configuration.get('purgeOnStart');
+
+        if (configuration.get('testPayPerEvent') === true) {
+            if (this.isAtHome) {
+                throw new Error('Using the ACTOR_TEST_PAY_PER_EVENT environment variable is only supported in a local development environment');
+            }
+
+            this.pricingModel = 'PAY_PER_EVENT';
+        }
+
+        this.apifyClient = apifyClient;
+    }
+
+    private get isPayPerEvent() {
+        return this.pricingModel === 'PAY_PER_EVENT';
+    }
+
+    /**
+     * Initialize the ChargingManager by loading pricing information and charging state via Apify API.
+     */
+    async init(): Promise<void> {
+        this.chargingState = {};
+
+        // Retrieve pricing information
+        if (this.isAtHome) {
+            const run = (await this.apifyClient.run(this.actorRunId!).get())!;
+            this.pricingModel = run.pricingInfo?.pricingModel;
+
+            // Load per-event pricing information
+            if (run.pricingInfo?.pricingModel === 'PAY_PER_EVENT') {
+                for (const [eventName, eventPricing] of Object.entries(run.pricingInfo.pricingPerEvent.actorChargeEvents)) {
+                    this.pricingInfo[eventName] = {
+                        price: eventPricing.eventPriceUsd,
+                        title: eventPricing.eventTitle,
+                    };
+                }
+
+                this.maxTotalChargeUsd = run.options.maxTotalChargeUsd ?? this.maxTotalChargeUsd;
+            }
+
+            // Load charged event counts
+            for (const [eventName, chargeCount] of Object.entries(run.chargedEventCounts ?? {})) {
+                this.chargingState[eventName] = {
+                    chargeCount,
+                    totalChargedAmount: chargeCount * (this.pricingInfo[eventName]?.price ?? 0),
+                };
+            }
+        }
+
+        if (this.isPayPerEvent) {
+            // Set up charging log dataset
+            if (this.isAtHome) {
+                const datasetId = await (async () => {
+                    const defaultStore = await KeyValueStore.open();
+
+                    const storedDatasetId = await defaultStore.getValue<string>(this.PLATFORM_CHARGING_LOG_DATASET_ID_KEY);
+                    if (storedDatasetId !== null) {
+                        return storedDatasetId;
+                    }
+
+                    const dataset = await this.apifyClient.datasets().getOrCreate();
```

Comment:
Q: Are you always creating this dataset? Seems you are. If that's the case, maybe this should be part of the platform itself? Who owns this dataset and is charged for its storage? How will it work with scoped tokens in the future?